framework/src/physics/PhysicsBase.C

```diff
@@ -477,3 +480,206 @@ PhysicsBase::addPetscPairsToPetscOptions(
         *this,
         po);
 }
+
+bool
+PhysicsBase::isVariableFV(const VariableName & var_name) const
+{
+  const auto var = &_problem->getVariable(0, var_name);
+  if (const auto & field_var = dynamic_cast<MooseVariableFieldBase *>(var))
+    return field_var->isFV();
+  return false;
+}
+
+bool
+PhysicsBase::isVariableScalar(const VariableName & var_name) const
+{
+  return _problem->hasScalarVariable(var_name);
+}
+
+bool
+PhysicsBase::shouldCreateVariable(const VariableName & var_name,
+                                  const std::vector<SubdomainName> & blocks,
+                                  const bool error_if_aux,
+                                  const bool add_missing_subdomains)
+{
+  if (!variableExists(var_name, error_if_aux))
+    return true;
+  // check block restriction
+  auto & var = _problem->getVariable(0, var_name);
+  if (var.hasBlocks(blocks))
+    return false;
+  else if (add_missing_subdomains)
+  {
+    var.addBlocks(blocks);
+    return false;
+  }
+  else
+    mooseError("Variable '" + var_name + "' already exists with subdomain restriction '" +
+               Moose::stringify(var.blocks()) + "' which does not include the subdomains '" +
+               Moose::stringify(blocks) + "', required for this Physics.");
+}
+
+bool
+PhysicsBase::shouldCreateIC(const VariableName & var_name,
+                            const std::vector<SubdomainName> & blocks,
+                            const bool ic_is_default_ic,
+                            const bool error_if_already_defined) const
+{
+  // Handle recover
+  if (ic_is_default_ic && (_app.isRestarting() || _app.isRecovering()))
+    return false;
+  // do not set initial conditions if we are loading fields from the mesh file
+  if (getParam<bool>("initialize_variables_from_mesh_file"))
+    return false;
+  // Different type of ICs, not block restrictable
+  mooseAssert(!isVariableScalar(var_name), "shouldCreateIC not implemented for scalar variables");
+
+  // Check whether there are any ICs for this variable already in the problem
+  bool has_all_blocks;
+  std::set<SubdomainName> blocks_covered;
+  const std::set<SubdomainName> blocks_set(blocks.begin(), blocks.end());
+  if (isVariableFV(var_name))
+    has_all_blocks = _problem->getFVInitialConditionWarehouse().hasObjectsForVariableAndBlocks(
+        var_name, blocks_set, blocks_covered, 0);
+  else
+    has_all_blocks = _problem->getInitialConditionWarehouse().hasObjectsForVariableAndBlocks(
```

Comment:
so as you cant see here and below, using the dedicated warehouse makes the code a lot smaller and readable than the query

the problem is that it adds to the dedicated warehouses, instead of developing theWarehouse() with new queries, which maybe we are not trying to do?

Reply:
would love some guidance on this